aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -42,383 +47,496 @@ module aptos_framework::fungible_asset {
 
     /// Transfer `ammount` of  fungible metadata with `TransferRef` even ungated transfer is disabled.
-    public fun transfer_with_ref(
+    public fun transfer_with_ref<T: key>(
         transfer_ref: &TransferRef,
-        from: &Object<FungibleAssetWallet>,
-        to: &Object<FungibleAssetWallet>,
+        from_wallet: Object<T>,
         amount: u64,
-    ) acquires FungibleAssetWallet {
-        let fa = withdraw_with_ref(transfer_ref, from, amount);
-        deposit_with_ref(transfer_ref, to, fa);
+        to_wallet: Object<T>,
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let fa = withdraw_with_ref(transfer_ref, from_wallet, amount);
+        deposit_with_ref(transfer_ref, to_wallet, fa);
     }
 
-    /// Get the underlying metadata object from `MintRef`.
-    public fun mint_ref_metadata(ref: &MintRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// Extract a given amount from the given fungible asset and return a new one.
+    public fun extract(fungible_asset: &mut FungibleAsset, amount: u64): FungibleAsset {
+        assert!(fungible_asset.amount >= amount, error::invalid_argument(EINSUFFICIENT_BALANCE));
+        fungible_asset.amount = fungible_asset.amount - amount;
+        FungibleAsset {
+            metadata: fungible_asset.metadata,
+            amount,
+        }
     }
 
-    /// Get the underlying metadata object from `TransferRef`.
-    public fun transfer_ref_metadata(ref: &TransferRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// Extract everything from the given fungible asset and return a new one.
+    public fun extract_all(fungible_asset: &mut FungibleAsset): FungibleAsset {
+        let amount = fungible_asset.amount;
+        fungible_asset.amount = 0;
+        FungibleAsset {
+            metadata: fungible_asset.metadata,
+            amount,
+        }
     }
 
-    /// Get the underlying metadata object from `BurnRef`.
-    public fun burn_ref_metadata(ref: &BurnRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// "Merges" the two given fungible assets. The coin passed in as `dst_fungible_asset` will have a value equal
+    /// to the sum of the two (`dst_fungible_asset` and `src_fungible_asset`).
+    public fun merge(dst_fungible_asset: &mut FungibleAsset, src_fungible_asset: FungibleAsset) {
+        let FungibleAsset { metadata: _, amount } = src_fungible_asset;
+        dst_fungible_asset.amount = dst_fungible_asset.amount + amount;
     }
 
-    /// Assert the owner of `metadata`.
-    inline fun assert_owner<T: key>(owner: &signer, metadata: &Object<T>) {
-        assert!(object::is_owner(*metadata, signer::address_of(owner)), error::permission_denied(ENOT_OWNER));
+    /// Destroy an empty fungible asset.
+    public fun destroy_zero(fungible_asset: FungibleAsset) {
+        let FungibleAsset { amount, metadata: _ } = fungible_asset;
+        assert!(amount == 0, error::invalid_argument(EAMOUNT_IS_NOT_ZERO));
+    }
```

Comment:
`destroy_zero` is not useful without `zero`. also remove?

Reply:
No. It could happen if you extract enough from an existing fungible asset. Since FungibleAsset is not droppable, having remaining 0 amount is annoying

Reply:
I'll remove extract_all. I think it's still useful to be able to destroy zero fungible asset in case users use extract somehow and there's nothing left.

Reply:
My thought is if users remove all funds, they will get a FA with that amount and leave an zero one. And you have to destroy the zero one. Instead in this case, you can just send the original one to avoid destroying. `extract_all` really doesn't make sense to me. And if you pass 0-balance FA around, if will only trigger the `AMOUNT_CANNOT_BE_ZERO` everywhere. So we just need to ensure FA with 0 amount will never exist across the framework to make our life easier.

Reply:
How do you ensure that it doesn't exist though? People could be extracting different amounts from the fungible assets (without checking the value) and leave a zero-amount at the end. extract_all makes sense to be removed but I'll leave destroy_zero just in case.